### PR TITLE
Member enrollment organization affiliation

### LINKED ISSUE #546: Add event name referral dropdown to member onboard enrollment
First issue draft: In the member onboarding enrollment sign-up form. Add a dropdown select input for event_names (ask Alan) and a text input for "Other". 

New issue draft: Want to know what organizations new signup members are affiliated with.

### Member 'contact us' quick-enroll and regular signup forms:
Label: Who are you affiliated with?
Multi-select field: Pre-populated dropdown list to choose from multiple organizations: 
Also list a "optional" freeform typed-in field.

Give admins the ability to update these selections in the backend.

### PR DESCRIPTION
**Pull request by DeeTheDev**
FIxes #546 

It would be helpful for suma staff to have a way to know what affiliated organizations members are eligible for when they signup. This change models Organizations and OrganizationMemberships, the organizations represent entities like "Hacienda CDC" or affordable housing partners, the membership represents the member who is affiliated with that organization.

When members signup regularly or through the contact list add page, we show them available organizations, they can choose one, we create a membership based on the organization chosen. If an organization does not exist, we allow them to type their own and we add a summary with that org so that suma staff can verify it in the adminapp. This helps expedite the member signup process for suma staff.

- Add model migrations and ensure that Organization Memberships are unique and not duplicated
- Add Organization & Organization Membership fixtures
- Setup adminapp CRUD routes for `Organization` and `Organization Memberships`
- Add ability to update/create/remove member organization memberships in the `MemberEditPage`
- Render memberships in the `MemberDetailPage`

### Onboarding Signup organization list input and dropdown choice
Screenshot 2024-04-11 at 8 00 48 PM

### Dropdown choice
Screenshot 2024-04-11 at 8 01 20 PM
Screenshot 2024-04-11 at 8 00 57 PM

### Same with Contact list add page
Screenshot 2024-04-11 at 8 02 20 PM

### Adminapp org and org membership lists
Screenshot 2024-04-11 at 8 06 26 PM
Screenshot 2024-04-11 at 8 05 00 PM

### Adminapp org and membership detail pages
Screenshot 2024-04-11 at 8 11 42 PM
Screenshot 2024-04-11 at 8 11 33 PM

### Summary from contact list add with optional organization 
Screenshot 2024-04-11 at 8 05 30 PM

### Member regular signup optional org chosen summary
Screenshot 2024-04-11 at 8 13 12 PM

### Member detail page org memberships
Screenshot 2024-04-15 at 7 41 56 PM

### Member Edit page - Dividers added, Memberships editable
Screenshot 2024-04-15 at 7 42 33 PM

### CODE REVIEW

File: db/migrations/045_member_organization.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+      text :name, null: false, unique: true
+    end
+
+    create_table(:organization_memberships) do
+      primary_key :id
+      timestamptz :created_at, null: false, default: Sequel.function(:now)
+      timestamptz :updated_at
+
+      foreign_key :member_id, :members, null: false, index: true
+      foreign_key :organization_id, :organizations, null: false, index: true, unique: true
```
Comment (rgalanakis) [inline db/migrations/045_member_organization.rb:18]: Did you try out creating multiple people within the same org?

File: db/migrations/045_member_organization.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+      timestamptz :updated_at
+      text :name, null: false, unique: true
+    end
+
+    create_table(:organization_memberships) do
+      primary_key :id
+      timestamptz :created_at, null: false, default: Sequel.function(:now)
+      timestamptz :updated_at
+
+      foreign_key :member_id, :members, null: false, index: true
```
Comment (rgalanakis) [inline db/migrations/045_member_organization.rb:17]: Use a new `index [:member_id, :organization_id], unique: true`. This needs a compound index, there should only be a single membership for a particular member in a particular org.

File: lib/suma/admin_api/members.rb [hunk 3]
```diff
@@ -216,6 +217,23 @@ def lookup_member!
...
         present member, with: DetailedMemberEntity
       end
+
+      params do
+        requires :membership_ids, type: Array[Integer], coerce_with: CommaSepArray[Integer]
+      end
+      delete :remove_memberships do
+        member = lookup_member!
+        params[:membership_ids].each do |id|
+          Suma::Organization::Membership[id] or adminerror!(403, "Unknown organization membership: #{id}")
```
Comment (rgalanakis) [inline lib/suma/admin_api/members.rb:227]: Why is this out here instead of in the transaction/same loop?

File: lib/suma/admin_api/search.rb [hunk 4]
```diff
@@ -8,6 +8,14 @@ class Suma::AdminAPI::Search < Suma::AdminAPI::V1
   include Suma::AdminAPI::Entities
 
   resource :search do
+    helpers do
+      def ds_search_or_order_by_name(ds, params)
```
Comment (rgalanakis) [inline lib/suma/admin_api/search.rb:12]: This is one of those things where the code is the same out of coincidence- if you rename something to not search on `name` this method would have to be rewritten. Instead let's refactor this to not be brittle/coincidentally DRY.

File: lib/suma/api/me.rb [hunk 5]
```diff
@@ -29,16 +29,28 @@ class Suma::API::Me < Suma::API::V1
       optional :address, type: JSON do
         use :address
       end
+      optional :organization_name, type: String, allow_blank: false
```
Comment (rgalanakis) [inline lib/suma/api/me.rb:32]: What does it mean for a member to update their organization at any time like this?

File: lib/suma/fixtures/organizations.rb [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  base :organization do
+    self.name ||= Faker::Company.name
+  end
+
+  before_saving do |instance|
+    instance
+  end
+
+  decorator :with_membership, presave: true do |m={}|
+    m = Suma::Fixtures.organization_membership.create(m) unless
```
Comment (rgalanakis) [inline lib/suma/fixtures/organizations.rb:21]: Think about what this is doing.

File: lib/suma/member.rb [hunk 7]
```diff
@@ -104,6 +105,10 @@ def with_email(*emails)
     def with_normalized_phone(*phones)
       return self.where(phone: phones)
     end
+
+    def active
```
Comment (rgalanakis) [inline lib/suma/member.rb:109]: There is already `not_soft_deleted` which is more explicit. What is 'active'?

File: lib/suma/organization.rb [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+require "suma/postgres/model"
+require "suma/admin_linked"
+
+class Suma::Organization < Suma::Postgres::Model(:organizations)
+  include Suma::AdminLinked
+  plugin :timestamps
+
+  one_to_many :memberships, class: "Suma::Organization::Membership"
+
+  def self.supported_organizations = self.all.map(&:name)
```
Comment (rgalanakis) [inline lib/suma/organization.rb:12]: This is not useful.

File: lib/suma/member.rb [hunk 9]
```diff
@@ -176,6 +181,12 @@ def replace_roles(roles)
     to_remove.each { |c| self.remove_role(c) }
   end
 
+  def create_organization_membership(organization_name)
+    organization = Suma::Organization[name: organization_name]
```
Comment (rgalanakis) [inline lib/suma/member.rb:185]: Don't do this- just pass in the organization. This whole method is unnecessary- the caller should have an org, then it's just `m.add_organization_membership(organization: org)`.

File: spec/suma/admin_api/members_spec.rb [hunk 10]
```diff
@@ -234,4 +234,25 @@ def make_item(i)
       expect(last_response).to have_status(403)
     end
   end
+
+  describe "DELETE /v1/members/:id/memberships" do
```
Comment (rgalanakis) [inline spec/suma/admin_api/members_spec.rb:238]: Let's figure out how we can bulk create/update/delete memberships rather than the endpoints here.

**Review by rgalanakis — CHANGES_REQUESTED** [review]

File: lib/suma/member.rb (hunk 7 above)
Comment (DeeTheDev) [reply lib/suma/member.rb:109]: Good catch, wasn't aware of it.

**Pull request merged**